src/main/kotlin/io/bazel/kotlin/builder/tasks/KotlinBuilder.kt

```diff
@@ -318,4 +347,266 @@ class KotlinBuilder
+
+      // Create temporary directories for KSP2 processing
+      val kspWorkDir = workingDir.resolve("_ksp2/$moduleName")
+      val stagedSourcesDir = kspWorkDir.resolve("staged_sources")
+      val kotlinOutputDir = kspWorkDir.resolve("kotlin_out")
+      val javaOutputDir = kspWorkDir.resolve("java_out")
+      val classOutputDir = kspWorkDir.resolve("class_out")
+      val resourceOutputDir = kspWorkDir.resolve("resource_out")
+      val cachesDir = kspWorkDir.resolve("caches")
+
+      listOf(
+        stagedSourcesDir,
+        kotlinOutputDir,
+        javaOutputDir,
+        classOutputDir,
+        resourceOutputDir,
+        cachesDir,
+      ).forEach {
+        Files.createDirectories(it)
+      }
+
+      try {
+        // Stage source files to isolated directory
+        val sourceRoots = mutableSetOf<String>()
+        val javaSourceRoots = mutableSetOf<String>()
+
+        // Stage individual source files
+        val sources = argMap.optional(KotlinBuilderFlags.KSP2_SOURCES) ?: emptyList()
+        for (source in sources) {
+          val sourceFile = File(source)
+          val targetFile = stagedSourcesDir.resolve(source).toFile()
+          targetFile.parentFile?.mkdirs()
+          sourceFile.copyTo(targetFile, overwrite = true)
+
+          // Track source roots (directories containing sources)
+          val sourceRoot =
+            if (sourceFile.parentFile != null) {
+              stagedSourcesDir.resolve(sourceFile.parentFile.path).toString()
+            } else {
+              stagedSourcesDir.toString()
+            }
+          sourceRoots.add(sourceRoot)
+          if (source.endsWith(".java")) {
+            javaSourceRoots.add(sourceRoot)
+          }
+        }
+
+        // Unpack srcjars directly
+        val srcjars = argMap.optional(KotlinBuilderFlags.KSP2_SOURCE_JARS) ?: emptyList()
+        for (srcjar in srcjars) {
+          ZipFile(srcjar).use { zip ->
+            zip.entries().asSequence().forEach { entry ->
+              if (!entry.isDirectory) {
+                val targetFile = stagedSourcesDir.resolve(entry.name).toFile()
+                targetFile.parentFile?.mkdirs()
+                zip.getInputStream(entry).use { input ->
+                  targetFile.outputStream().use { output ->
+                    input.copyTo(output)
+                  }
+                }
+                // Track source root for srcjar contents
+                val parentDir = targetFile.parentFile?.path ?: stagedSourcesDir.toString()
+                sourceRoots.add(parentDir)
+                if (entry.name.endsWith(".java")) {
+                  javaSourceRoots.add(parentDir)
+                }
+              }
+            }
+          }
+        }
+
+        // If no sources, add a placeholder source root
+        if (sourceRoots.isEmpty()) {
+          sourceRoots.add(stagedSourcesDir.toString())
+        }
+
+        // Load KSP2 via reflection
```

Comment:
It's a usually saner to use reflection once, ala [BuildToolsAPICompiler](https://github.com/bazelbuild/rules_kotlin/blob/master/src/main/kotlin/io/bazel/kotlin/compiler/BuildToolsAPICompiler.kt) than to use reflect for every class.

Reflection has overhead on the worker runtime, so we try to use it as little as possible.